src/KeyboardMoveResizeEventHandler.hh

```diff
@@ -130,103 +130,98 @@ public:
+            gm_mask = Y_VALUE;
             break;
         case RESIZE_HORIZONTAL:
             if (action.getParamI(1) == UNIT_PERCENT) {
                 Geometry head;
-                X11::getHeadInfo(X11Util::getNearestHead(*_decor), head);
+                X11::getHeadInfo(X11Util::getNearestHead(*_frame), head);
                 _gm.width += (action.getParamI(0)
                               * static_cast<int>(head.width)) / 100;
             } else {
                 _gm.width +=  action.getParamI(0);
             }
-            if (! _outline) {
-                _decor->resize(_gm.width, _gm.height);
-            }
+            gm_mask = WIDTH_VALUE;
             break;
         case RESIZE_VERTICAL:
             if (action.getParamI(1) == UNIT_PERCENT) {
                 Geometry head;
-                X11::getHeadInfo(X11Util::getNearestHead(*_decor), head);
+                X11::getHeadInfo(X11Util::getNearestHead(*_frame), head);
                 _gm.height += (action.getParamI(0)
                                * static_cast<int>(head.height)) / 100;
             } else {
                 _gm.height +=  action.getParamI(0);
             }
-            if (! _outline) {
-                _decor->resize(_gm.width, _gm.height);
-            }
+            gm_mask = HEIGHT_VALUE;
+            break;
+        case MOVE_SET_GEOMETRY:
+            _frame->prepareSetGeometry(_gm, gm_mask,
+                                       action.getParamS(),
+                                       action.getParamI(0),
+                                       action.getParamI(1) == 1);
             break;
         case MOVE_SNAP:
-            PDecor::checkSnap(_decor, _gm);
-            if (! _outline) {
-                _decor->move(_gm.x, _gm.y);
-            }
+            PDecor::checkSnap(_frame, _gm);
+            gm_mask = X_VALUE | Y_VALUE;
             break;
         case MOVE_CANCEL:
+            gm_mask = Frame::geometryMask(_old_gm, _gm);
             _gm = _old_gm; // restore position
 
             if (! _outline) {
-                _decor->move(_gm.x, _gm.y);
-                _decor->resize(_gm.width, _gm.height);
+                _frame->setGeometry(_old_gm, gm_mask);
             }
 
             return EventHandler::EVENT_STOP_PROCESSED;
         case MOVE_END:
             if (_outline) {
-                if (_gm.x != _old_gm.x
-                    || _gm.y != _old_gm.y) {
-                    _decor->move(_gm.x, _gm.y);
-                }
-                if (_gm.width != _old_gm.width
-                    || _gm.height != _old_gm.height) {
-                    _decor->resize(_gm.width, _gm.height);
-                }
+                gm_mask = Frame::geometryMask(_gm, _old_gm);
+                _frame->setGeometry(_gm, gm_mask);
             }
             return EventHandler::EVENT_STOP_PROCESSED;
         default:
             break;
         }
 
+        if (! _outline) {
+            _frame->setGeometry(_gm, gm_mask);
```

Comment:
Any reason why setGeometry could not be moved to PDecor instead of having it in Frame?

Reply:
Agh.. I missed this. Will do next, yes i think it's possible but some more vars may have to be moved too.